akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala

```diff
@@ -945,35 +946,92 @@ private[cluster] class ClusterCoreDaemon(publisher: ActorRef) extends Actor with
+   * means it is a choice among all nodes of the cluster.
+   */
+  def localDcGossipTargets(): Vector[UniqueAddress] = {
+    val firstSelection =
+      if (ThreadLocalRandom.current.nextDouble() < adjustedGossipDifferentViewProbability) {
+        // If it's time to try to gossip to some nodes with a different view
+        // gossip to a random alive same dc member with preference to a member with older gossip version
+        latestGossip.members.collect {
+          case m if m.dataCenter == selfDc && !latestGossip.seenByNode(m.uniqueAddress) && validNodeForGossip(m.uniqueAddress) ⇒
+            m.uniqueAddress
+        }(breakOut)
+      } else Vector.empty
+
+    // Fall back to localGossip
+    if (firstSelection.isEmpty) {
+      latestGossip.members.toVector.collect {
+        case m if m.dataCenter == selfDc && validNodeForGossip(m.uniqueAddress) ⇒ m.uniqueAddress
+      }
+    } else Vector.empty
+
+  }
+
+  /**
+   * Choose cross-dc nodes if this one of the N oldest nodes, and if not fall back to gosip locally in the dc
+   */
+  def multiDcGossipTargets(): Vector[UniqueAddress] = {
+
+    // 20% of the times across dcs and 80% local (doing it for all nodes to avoid having to collect the oldest nodes per
+    // dc all the time, when #23290 is merged we can maybe cache oldest members per dc in that)
+    if (ThreadLocalRandom.current.nextDouble() > 0.2D) localDcGossipTargets()
+    else {
+      // TODO cache this until gossip changes??
+      val nodesPerDc = latestGossip.members.foldLeft(Map.empty[String, SortedSet[Member]]) { (acc, member) ⇒
+        acc.get(member.dataCenter) match {
+          case Some(set) if set.size > cluster.settings.CrossDcConnections ⇒ acc
+          case Some(set) ⇒ acc + (member.dataCenter → (set + member))
+          case None ⇒ acc + (member.dataCenter → (SortedSet.empty(Member.ageOrdering) + member))
         }
       }
+
+      // only do cross DC gossip if this node is among the N oldest
+      val selfMember = latestGossip.member(selfUniqueAddress)
+      if (!nodesPerDc(selfDc).contains(selfMember)) localDcGossipTargets()
+      else {
+        def findFirstDcWithValidNodes(left: List[DataCenter], onlyUnseen: Boolean): Vector[UniqueAddress] =
+          left match {
+            case dc :: tail ⇒
+              val validNodes = nodesPerDc(dc).collect {
+                case member if validNodeForGossip(member.uniqueAddress) && (!onlyUnseen || latestGossip.seenByNode(member.uniqueAddress)) ⇒
+                  member.uniqueAddress
+              }
+              if (validNodes.nonEmpty) validNodes.toVector
+              else findFirstDcWithValidNodes(tail, onlyUnseen) // no valid nodes in dc, try next
+
+            case Nil ⇒
+              Vector.empty
+          }
+
+        // chose another DC at random
+        val otherDcsInRandomOrder = Random.shuffle((nodesPerDc - selfDc).keys.toList)
```

Comment:
You do a lot of stuff with the sorting and finding nodes in other DCs than the one you finally select.
You can select DC first, and then only find X oldest for that DC.